source/mainnet/smart-contracts/guides/unit-test-contract.rst

```diff
@@ -352,6 +352,85 @@ state and balance fields:
       // *host.state() and state_copy might not be equal any more due to reentrancy.
       do_something_with(state_copy);
 
+Testing with state rollbacks
+============================
+
+Invocations of smart contracts on the chain are transactional. This means that
+if a contract changes its state and then fails, the state is rolled back to how
+it was before the invocation.
+
+If we want the same behavior when testing, it is necessary to use a helper
+method on the |TestHost|_, namely |with_rollback|_.
+To illustrate, here is an example in which the receive function increments the
+state and then immediately fails:
+
+.. code-block:: rust
+   :emphasize-lines: 23, 25, 35, 37
+
+   type State = u8;
+
+   #[receive(contract = "my_contract", name = "increment", mutable)]
+   fn receive<S: HasStateApi>(
+       _ctx: &impl HasReceiveContext,
+       host: &mut impl HasHost<State, StateApiType = S>,
+   ) -> ReceiveResult<()> {
+       *host.state_mut() += 1; // Mutate state.
+       Err(Reject::default())  // Then fail.
+   }
+
+   #[concordium_cfg_test]
+   mod tests {
+       use super::*;
+       use concordium_std::test_infrastructure::*;
+
+       #[test]
+       fn test_without_rollback() {
+           let state = 0;
+           let ctx = TestReceiveContext::empty();
+           let mut host = TestHost::new(state, StateBuilder::new());
+
+           let _ = receive(&ctx, &mut host);
+
+           claim_eq!(*host.state(), 0); // FAILS! State wasn't rolled back.
+       }
+
+       #[test]
+       fn test_with_rollback() {
+           let state = 0;
+           let ctx = TestReceiveContext::empty();
+           let mut host = TestHost::new(state, StateBuilder::new());
+
+           // Use the `with_rollback` method.
+           let _ = host.with_rollback(|host| receive(&ctx, host));
+
+           claim_eq!(*host.state(), 0); // Success!
+       }
+   }
+
+|with_rollback|_ works by creating a clone of the ``State``, invoking the
+receive function and, if it failed, rolling back the state.
+This means that ``State`` must implement the trait |StateClone|_, which
+fortunately is implemented for all |Clone|_ types.
+If your ``State`` has one or more fields comprised of |StateBox|_, |StateSet|_, or |StateMap|_, it is
+necessary to derive |StateClone|_ directly. For example:
```

Comment:
It would be useful if we were a bit more direct here and stated that
- either your state does not use any StateBox/StateSet/StateMap in which case you should derive `Clone` for it
- if it does, then you should derive `StateClone` for it.